src/magic_spell_effect.cpp

```diff
@@ -504,8 +504,11 @@ static void magical_polymorph( monster &victim, Creature &caster, const spell &s
                 }
                 if( ( mtypes[iter].id != victim.type->id ) && ( std::abs( mtypes[iter].difficulty - victim_diff )
                         <= difficulty_variance ) ) {
-                    new_id = mtypes[iter].id;
-                    break;
+                    if( !mtypes[iter].in_species( species_id( "HALLUCINATION" ) ) &&
+                        mtypes[iter].id.str() != "mon_generator" ) {
```

Comment:
```suggestion
                        mtypes[iter].id != mtype_id( "mon_generator" ) ) {
```
this _should_ be faster than just a string compare iirc

Reply:
or, well maybe it should be defined as a static variable

Reply:
This will be slower, not faster, but I don't think the speed is relevant and it's clearer this way, so I'd still support the change..